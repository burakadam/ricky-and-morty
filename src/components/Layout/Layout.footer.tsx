import { Typography } from '@mui/material';
import { Link } from 'react-router-dom';

import IMAGE from '@/assets/images/rick-and-morty-green-hole.png';
import ROUTES from '@/constants/routes';

import * as S from './Layout.styles';

const Footer = () => (
  <S.Footer>
    <Link to={ROUTES.CHARACTERS}>
      <S.FooterImage src={IMAGE} />
    </Link>
    <S.CreatedBy>
      <Typography variant="caption" display="block">
        Created By:
      </Typography>
      <Typography variant="caption" display="block" color="crimson" fontWeight="bold">
        Dan Harmon - Justin Roiland
      </Typography>
    </S.CreatedBy>
  </S.Footer>
);

export { Footer };
