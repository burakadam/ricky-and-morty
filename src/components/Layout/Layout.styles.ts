import { AppBar } from '@mui/material';
import { styled } from '@mui/system';

const Container = styled('div')`
  padding: 50px 20px 0 20px;
`;

const Content = styled('div')`
  min-height: 100vh;
  width: 100%;
  max-width: 1024px;
  margin: auto;
  padding: 30px 0;
`;

const Header = styled(AppBar)`
  height: 50px;
  flex-direction: row;
  justify-content: space-between;
  padding: 0 10px;
`;

const Logo = styled('img')`
  height: 40px;
  width: auto;
  align-self: flex-start;
  margin-top: 5px;
`;

const TextLogo = styled('img')`
  height: 30px;
  width: auto;
  align-items: flex-end;
  margin-top: 10px;
`;

const Footer = styled('footer')`
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 0 10px 20px 10px;
`;

const FooterImage = styled('img')`
  width: 100%;
  max-width: 150px;
`;

const CreatedBy = styled('div')`
  display: flex;
  margin: 20px 0;
  opacity: 0.8;
  > span {
    margin: 0 0.2em;
  }
`;

export { Container, Content, CreatedBy, Footer, FooterImage, Header, Logo, TextLogo };
