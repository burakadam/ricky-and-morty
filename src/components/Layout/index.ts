import type React from 'react';

export interface IHideOnScrollProps {
  children: React.ReactElement;
}

export * from './Layout';
