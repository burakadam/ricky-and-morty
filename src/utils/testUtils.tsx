import type { PreloadedState } from '@reduxjs/toolkit';
import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import type { PropsWithChildren } from 'react';
import React from 'react';
import { Provider } from 'react-redux';

import type { AppStore, RootState } from '@/store/store';
import { setupStore } from '@/store/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  // eslint-disable-next-line @typescript-eslint/no-shadow
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function Wrapper({ children }: { children: React.ReactElement }) {
  return <>{children}</>;
}

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
