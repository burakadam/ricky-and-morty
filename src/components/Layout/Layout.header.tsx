import { Slide, useScrollTrigger } from '@mui/material';

import TEXT_LOGO from '@/assets/images/rick-and-morty-logo-text.png';
import LOGO from '@/assets/images/rick-and-morty-silhouette.jpg';

import type { IHideOnScrollProps } from '.';
import * as S from './Layout.styles';

function HideOnScroll(props: IHideOnScrollProps) {
  const trigger = useScrollTrigger({ target: window || undefined });

  return (
    <Slide appear={false} direction="down" in={!trigger}>
      {props.children}
    </Slide>
  );
}

const Header = () => (
  <HideOnScroll>
    <S.Header position="fixed" color="primary">
      <S.Logo src={LOGO} />
      <S.TextLogo src={TEXT_LOGO} />
    </S.Header>
  </HideOnScroll>
);

export { Header };
