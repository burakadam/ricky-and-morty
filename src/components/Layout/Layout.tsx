import { useEffect } from 'react';
import { Outlet, useLocation } from 'react-router-dom';

import { Footer } from './Layout.footer';
import { Header } from './Layout.header';
import * as S from './Layout.styles';

const Layout = () => {
  const { pathname } = useLocation();

  useEffect(() => {
    document.documentElement.scrollTo({
      top: 0,
      left: 0,
      behavior: 'smooth'
    });
  }, [pathname]);

  return (
    <S.Container>
      <Header />
      <S.Content>
        <Outlet />
      </S.Content>
      <Footer />
    </S.Container>
  );
};

export { Layout };
