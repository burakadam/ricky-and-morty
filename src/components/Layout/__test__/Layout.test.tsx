// import { screen } from '@testing-library/react';
import { MemoryRouter as Router } from 'react-router-dom';

import { renderWithProviders, screen } from '@/utils/testUtils';

import { Layout } from '../Layout';

describe('App Layout component', () => {
  let container: HTMLElement;

  beforeEach(() => {
    // eslint-disable-next-line testing-library/no-render-in-setup
    ({ container } = renderWithProviders(
      <Router>
        <Layout />
      </Router>
    ));
  });

  it('render component correctly', () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it('header has a home url', () => {
    const link = container.querySelector('header a') as HTMLAnchorElement;
    expect(link.href).toEqual(`http://localhost/characters`);
  });

  it('footer has creaters names', () => {
    expect(screen.getByText('Dan Harmon - Justin Roiland')).toBeInTheDocument();
  });
});
