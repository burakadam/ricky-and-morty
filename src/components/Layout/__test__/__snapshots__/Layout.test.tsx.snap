// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Layout component render component correctly 1`] = `
<div
  class="css-14pq0uy"
>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1o5cd81-MuiPaper-root-MuiAppBar-root"
  >
    <a
      href="/characters"
    >
      <img
        alt="Rick and Morty silhoutte"
        class="css-1j19rp1"
        src="rick-and-morty-silhouette.jpg"
      />
    </a>
    <img
      alt="Rick and Morty show text logo"
      class="css-tbbnae"
      src="rick-and-morty-logo-text.png"
    />
  </header>
  <div
    class="css-r63hsx"
  />
  <footer
    class="css-1n9380"
  >
    <a
      href="/characters"
    >
      <img
        class="css-blwqou"
        src="rick-and-morty-green-hole.png"
      />
    </a>
    <div
      class="css-12voi5e"
    >
      <span
        class="MuiTypography-root MuiTypography-caption css-11idbvn-MuiTypography-root"
      >
        Created By:
      </span>
      <span
        class="MuiTypography-root MuiTypography-caption css-2rpwgx-MuiTypography-root"
      >
        Dan Harmon - Justin Roiland
      </span>
    </div>
  </footer>
</div>
`;
